add messages for anthropic sdk compatibility

dsingal0 opened the pull request:
<!--
  What does this PR add, remove, and/or change?
-->
## :rocket: What
Adds /v1/messages
<!--
  How was the change described above implemented?
-->
## :computer: How
<!--
  How have I ensured release and ongoing quality of this change?
-->
## :microscope: Testing
tested by pip installing truss from src, running openai_test with curl to curl -X POST http://localhost:8080/v1/messages \
  -H "Content-Type: application/json" \
  -d '{"key1": "value1", "key2": "value2"}'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/basetenlabs/truss?pullRequest=1883) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/basetenlabs/truss?pullRequest=1883) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: dsingal0<br/>:x: Dhruv Singal<br/><hr/>**Dhruv Singal** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/basetenlabs/truss?pullRequest=1883) it.</sub>

dsingal0 commented:
Refreshed this old PR on top of current \ and expanded the compatibility work to include \, \, and \.

Verification run locally:
- \...:1:1: E902 No such file or directory (os error 2)
Found 1 error.
- \============================= test session starts ==============================
platform darwin -- Python 3.13.3, pytest-7.2.0, pluggy-1.6.0 -- /private/tmp/truss-pr1883/.venv/bin/python
cachedir: .pytest_cache
rootdir: /private/tmp/truss-pr1883, configfile: pyproject.toml
plugins: asyncio-0.23.8, check-2.5.3, anyio-4.10.0, split-0.10.0, cov-3.0.0, requests-mock-1.12.1
asyncio: mode=Mode.STRICT
collecting ... collected 7 items

truss/tests/templates/server/test_model_wrapper.py::test_model_wrapper_load_error_once PASSED [ 14%]
truss/tests/templates/server/test_model_wrapper.py::test_model_wrapper_load_error_more_than_allowed PASSED [ 28%]
truss/tests/templates/server/test_model_wrapper.py::test_model_wrapper_streaming_timeout PASSED [ 42%]
truss/tests/templates/server/test_model_wrapper.py::test_trt_llm_truss_predict PASSED [ 57%]
truss/tests/templates/server/test_model_wrapper.py::test_trt_llm_truss_missing_model_py PASSED [ 71%]
truss/tests/templates/server/test_model_wrapper.py::test_open_ai_completion_endpoints PASSED [ 85%]
truss/tests/templates/server/test_model_wrapper.py::test_messages_endpoint PASSED [100%]

=============================== warnings summary ===============================
.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:940: 17 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:940: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    inlocs = ast.Compare(ast.Str(name.id), [ast.In()], [locs])

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:943: 16 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:943: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expr = ast.IfExp(test, self.display(name), ast.Str(name.id))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1053: 11 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1053: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    syms.append(ast.Str(sym))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1055: 11 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1055: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expls.append(ast.Str(expl))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:804: 46 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:804: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    keys = [ast.Str(key) for key in current.keys()]

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:914: 14 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:914: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    assertmsg = ast.Str("")

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:916: 14 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:916: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    template = ast.BinOp(assertmsg, ast.Add(), ast.Str(explanation))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:928: 13 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:928: DeprecationWarning: ast.NameConstant is deprecated and will be removed in Python 3.14; use ast.Constant instead
    clear = ast.Assign(variables, ast.NameConstant(None))

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================= 7 passed, 142 warnings in 3.60s ========================
- \============================= test session starts ==============================
platform darwin -- Python 3.13.3, pytest-7.2.0, pluggy-1.6.0 -- /private/tmp/truss-pr1883/.venv/bin/python
cachedir: .pytest_cache
rootdir: /private/tmp/truss-pr1883, configfile: pyproject.toml
plugins: asyncio-0.23.8, check-2.5.3, anyio-4.10.0, split-0.10.0, cov-3.0.0, requests-mock-1.12.1
asyncio: mode=Mode.STRICT
collecting ... collected 69 items / 68 deselected / 1 selected

truss/tests/test_model_inference.py::test_custom_openai_endpoints FAILED [100%]

=================================== FAILURES ===================================
_________________________ test_custom_openai_endpoints _________________________

    @contextmanager
    def ensure_kill_all():
        try:
>           with _show_container_logs_if_raised():

truss/tests/test_testing_utilities_for_other_tests.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <contextlib._GeneratorContextManager object at 0x10c8f4130>

    def __enter__(self):
        # do not keep args and kwds alive unnecessarily
        # they are only needed for recreation, which is not possible anymore
        del self.args, self.kwds, self.func
        try:
>           return next(self.gen)

/Users/dsingal/.local/share/uv/python/cpython-3.13.3-macos-aarch64-none/lib/python3.13/contextlib.py:141: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @contextmanager
    def _show_container_logs_if_raised():
>       initial_ids = {c.id for c in get_containers({TRUSS: True})}

truss/tests/test_testing_utilities_for_other_tests.py:43: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

labels = {'truss': True}, all = False

    def get_containers(labels: Dict, all: bool = False) -> list["Container"]:
        """Gets containers given labels."""
>       return Docker.client().container.list(
            filters=_create_label_filters(labels), all=all
        )

truss/util/docker.py:29: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <python_on_whales.components.container.cli_wrapper.ContainerCLI object at 0x10c9e9be0>
all = False, filters = dict_items([('label=truss', True)])

    def list(
        self,
        all: bool = False,
        filters: Union[Iterable[ContainerListFilter], Mapping[str, Any]] = (),
    ) -> List[Container]:
        """List the containers on the host.
    
        Alias: `docker.ps(...)`
    
        Parameters:
            all: If `True`, also returns containers that are not running.
            filters: Filters to apply when listing containers.
    
        # Returns
            A `List[python_on_whales.Container]`
        """
        if isinstance(filters, Mapping):
            filters = filters.items()
            warnings.warn(
                "Passing filters as a mapping is deprecated, replace with an "
                "iterable of tuples instead, as so:\n"
                f"filters={list(filters)}",
                DeprecationWarning,
            )
>       full_cmd = self.docker_cmd

.venv/lib/python3.13/site-packages/python_on_whales/components/container/cli_wrapper.py:1207: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <python_on_whales.components.container.cli_wrapper.ContainerCLI object at 0x10c9e9be0>

    @property
    def docker_cmd(self) -> Command:
>       return self.client_config.docker_cmd

.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ClientConfig(config=None, context=None, debug=None, host=None, log_level=None, tls=None, tlscacert=None, tlscert=None,..._directory=None, compose_compatibility=None, client_call=['docker'], client_type='docker', _client_call_with_path=None)

    @property
    def docker_cmd(self) -> Command:
>       result = Command(self.get_client_call_with_path())

.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:100: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ClientConfig(config=None, context=None, debug=None, host=None, log_level=None, tls=None, tlscacert=None, tlscert=None,..._directory=None, compose_compatibility=None, client_call=['docker'], client_type='docker', _client_call_with_path=None)

    def get_client_call_with_path(self) -> List[Union[Path, str]]:
        if self._client_call_with_path is None:
            self._client_call_with_path = [
>               Path(self._get_docker_path())
            ] + self.client_call[1:]

.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:83: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ClientConfig(config=None, context=None, debug=None, host=None, log_level=None, tls=None, tlscacert=None, tlscert=None,..._directory=None, compose_compatibility=None, client_call=['docker'], client_type='docker', _client_call_with_path=None)

    def _get_docker_path(self) -> str:
        which_result = shutil.which(self.client_call[0])
        if which_result is None:
>           raise ClientNotFoundError(
                f"The binary '{self.client_call[0]}' could not be found on your PATH. "
                "Please ensure that your PATH has the directory of the binary you're looking for. "
                "You can use `print(os.environ['PATH'])` to verify what directories are in your PATH."
            )
E           python_on_whales.client_config.ClientNotFoundError: The binary 'docker' could not be found on your PATH. Please ensure that your PATH has the directory of the binary you're looking for. You can use `print(os.environ['PATH'])` to verify what directories are in your PATH.

.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:91: ClientNotFoundError

During handling of the above exception, another exception occurred:

    @pytest.mark.integration
    def test_custom_openai_endpoints():
        """
        Test a Truss that exposes an OpenAI compatible endpoint.
        """
        model = """
        class Model:
            def load(self):
                self._predict_count = 0
                self._completions_count = 0
                self._embeddings_count = 0
                self._messages_count = 0
                self._responses_count = 0
    
            async def predict(self, inputs) -> int:
                self._predict_count += inputs["increment"]
                return self._predict_count
    
            async def completions(self, inputs) -> int:
                self._completions_count += inputs["increment"]
                return self._completions_count
    
            async def embeddings(self, inputs) -> int:
                self._embeddings_count += inputs["increment"]
                return self._embeddings_count
    
            async def messages(self, inputs) -> int:
                self._messages_count += inputs["increment"]
                return self._messages_count
    
            async def responses(self, inputs) -> int:
                self._responses_count += inputs["increment"]
                return self._responses_count
        """
>       with ensure_kill_all(), _temp_truss(model) as tr:

truss/tests/test_model_inference.py:1820: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/dsingal/.local/share/uv/python/cpython-3.13.3-macos-aarch64-none/lib/python3.13/contextlib.py:141: in __enter__
    return next(self.gen)
truss/tests/test_testing_utilities_for_other_tests.py:23: in ensure_kill_all
    kill_all_with_retries()
truss/tests/test_testing_utilities_for_other_tests.py:72: in kill_all_with_retries
    kill_all()
truss/util/docker.py:127: in kill_all
    kill_containers({TRUSS: True})
truss/util/docker.py:79: in kill_containers
    containers = get_containers(labels)
truss/util/docker.py:29: in get_containers
    return Docker.client().container.list(
.venv/lib/python3.13/site-packages/python_on_whales/components/container/cli_wrapper.py:1207: in list
    full_cmd = self.docker_cmd
.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:163: in docker_cmd
    return self.client_config.docker_cmd
.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:100: in docker_cmd
    result = Command(self.get_client_call_with_path())
.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:83: in get_client_call_with_path
    Path(self._get_docker_path())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ClientConfig(config=None, context=None, debug=None, host=None, log_level=None, tls=None, tlscacert=None, tlscert=None,..._directory=None, compose_compatibility=None, client_call=['docker'], client_type='docker', _client_call_with_path=None)

    def _get_docker_path(self) -> str:
        which_result = shutil.which(self.client_call[0])
        if which_result is None:
>           raise ClientNotFoundError(
                f"The binary '{self.client_call[0]}' could not be found on your PATH. "
                "Please ensure that your PATH has the directory of the binary you're looking for. "
                "You can use `print(os.environ['PATH'])` to verify what directories are in your PATH."
            )
E           python_on_whales.client_config.ClientNotFoundError: The binary 'docker' could not be found on your PATH. Please ensure that your PATH has the directory of the binary you're looking for. You can use `print(os.environ['PATH'])` to verify what directories are in your PATH.

.venv/lib/python3.13/site-packages/python_on_whales/client_config.py:91: ClientNotFoundError
=============================== warnings summary ===============================
.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:940: 218 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:940: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    inlocs = ast.Compare(ast.Str(name.id), [ast.In()], [locs])

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:943: 217 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:943: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expr = ast.IfExp(test, self.display(name), ast.Str(name.id))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1053: 235 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1053: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    syms.append(ast.Str(sym))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1055: 235 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:1055: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expls.append(ast.Str(expl))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:804: 1035 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:804: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    keys = [ast.Str(key) for key in current.keys()]

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:914: 233 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:914: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    assertmsg = ast.Str("")

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:916: 240 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:916: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    template = ast.BinOp(assertmsg, ast.Add(), ast.Str(explanation))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:928: 241 warnings
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:928: DeprecationWarning: ast.NameConstant is deprecated and will be removed in Python 3.14; use ast.Constant instead
    clear = ast.Assign(variables, ast.NameConstant(None))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:965
.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:965
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:965: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expl_format = self.pop_format_context(ast.Str(expl))

.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:977
.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:977
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/_pytest/assertion/rewrite.py:977: DeprecationWarning: ast.Num is deprecated and will be removed in Python 3.14; use ast.Constant instead
    expl_template = self.helper("_format_boolop", expl_list, ast.Num(is_or))

truss/tests/test_model_inference.py::test_custom_openai_endpoints
truss/tests/test_model_inference.py::test_custom_openai_endpoints
  /private/tmp/truss-pr1883/.venv/lib/python3.13/site-packages/python_on_whales/components/container/cli_wrapper.py:1201: DeprecationWarning: Passing filters as a mapping is deprecated, replace with an iterable of tuples instead, as so:
  filters=[('label=truss', True)]
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED truss/tests/test_model_inference.py::test_custom_openai_endpoints - py...
=============== 1 failed, 68 deselected, 2660 warnings in 0.29s ================ could not run here because \ is unavailable.

cretz commented on truss/templates/server/truss_server.py:
```diff
@@ -461,6 +482,24 @@ async def lifespan(app):
+                    tags=["V1"],
+                ),
+                FastAPIRoute(
+                    r"/v1/messages",
```
> Refresh this PR on top of current main and add OpenAI-compatible /v1/messages, /v1/responses, and /v1/embeddings endpoints for custom Truss models.

While I can find `/embeddings` and `/responses` in OpenAI API, I can't find `/messages`. I see there is some legacy stuff around `/threads/{thread_id}/messages`, is `/messages` an undocumented legacy endpoint? Or is this Anthropic compat?

dsingal0 replied:
/v1/messages is anthropic compat.

cretz replied:
👍 (may be worth PR title/description update, but no big deal, LGTM)
